Implement mod panel for new mod select design

bdach opened the pull request:
Continuation of #16917.

Reference design: https://www.figma.com/file/NkdREXr6wFkA99oLw8PnBv/Client%2FMod-Selection?node-id=0%3A17

The diff is a bit larger than I would have liked but most of this is UI code, somewhat artificially inflated by me moving out one tooltip class from the current implementation to share (which makes up for ~100 lines of the diffstat).

https://user-images.githubusercontent.com/20418176/155420531-ae401365-c922-4eee-ba92-344f98354fdd.mp4

List of changes / enhancements made to design that are very much up for discussion:

* Hover state has been changed to add accents to background of both parts of the panel rather than just the content panel as on design.
* Colours in active state have been inverted to preserve consistency (larger panel always darker, inner content panel always brighter)
* Hover + active state has been added (not there on designs at all).
* When the button is held down, a partial resize animation plays in order to better convey what is about to happen when the button is released.
* Check on/off click samples are used to match the old mod icons.
* An "incompatible" state for the panel has been implemented, in which case the panel is basically darkened and disabled. This is a change from the current mod select, in which the buttons will always be active, and clicking them will cause the other incompatible mod to be excluded from the set of active mods.
* Multi mods are not present on design in any way and so no consideration has been made here for that. The designs depict a flat list and that's the assumption I'm currently working with.
* All measurements from figma have been scaled down by a factor of 1.4 and rounded by eye to make numbers less ugly. The 1.4 number comes from the fact that the design is spec'd at 1920x1080, while the game has a `DrawSizePreservingFillContainer` with a size of 1366x768. 1920/1366 works out roughly to 1.4.

  Note that test scenes have their own `DrawSizePreservingFillContainer` and as such the real pixel measurements on screen will be ~5 pixels smaller than they will actually be in game.
* I'm aware that the shared incompatibility displaying tooltip has the mod description on it. I'm keeping it there for now until the old implementation goes away (?) and the description can finally be removed. May also be a decent fallback for cases where the description doesn't fully fit in the panel?

bdach committed:
Implement mod panel for new mod select screen

bdach committed:
Move out incompatibility displaying tooltip to own class

bdach committed:
Implement incompatibility-displaying variant of mod panel

peppy committed:
Merge branch 'master' into mod-overlay/panel

peppy commented on osu.Game/Overlays/Mods/ModPanel.cs:
```diff
@@ -0,0 +1,272 @@
+        protected override bool OnDragStart(DragStartEvent e)
+        {
+            mouseDownTime = null;
+            return true;
```
This leads to unexpected behaviour as a button (dragging away and then back should still activate the button). Seems like it was added to avoid cross-talk between the animation states, but that's not really how this should be used. I have some changes which restructure things a bit to avoid this being necessary.

peppy committed:
Combine and simplify state management logic

This makes a few changes to bring things into a better shape during
mouse interactions with the mod panels:

- Dragging away from the panel now works in line with other buttons (ie.
  `OsuButton`)
- Hovering now uses a lightened version of the current state, rather
  than always using the active colour. I think this feels better.
- Mouse down now uses a transform point of 0.5. This is to give the
  button a feeling of one of those latching light switches which resists
until reaching a point of overcoming the spring and switching state. I
think 0.4 (non-active) and 0.6 (from active) may work better, but left
at 0.5 for simplicity of implementation and I think it's good enough?
- Border always uses the gradiented version. I did this for simplicity
  of implementation, but also think it looks better.
- Adjusted transform durations to feel better to me.

peppy commented:
This PR makes use of `Colour4`, but I'm not sure how/when we should be using this. Was this an intentional usage for colour matching reasons?

bdach commented:
Not sure why the `Colour4` concern came up here and now, I've been using that since that was created. There should be no colour matching concerns there. I believe that was introduced to eventually facilitate decoupling from osutk's `Color4`.

peppy commented:
I guess we should look to be updating `ColourProvider` and other similar classes which still expose `Color4` and require casts in many places then.

peppy commented:
I've pushed some changes to state handling to tidy things up a touch, and alter the visual transforms in line with my own expectations.

https://user-images.githubusercontent.com/191335/155470490-98c98a42-b0ef-4f29-a9c1-ca692f7237ae.mp4

- Dragging away from the panel now works in line with other buttons (ie.  `OsuButton`)
- Hovering now uses a lightened version of the current state, rather than always using the active colour. I think this feels better.  
- Mouse down now uses a transform point of 0.5. This is to give the button a feeling of one of those latching light switches which resists until reaching a point of overcoming the spring and switching state. I think 0.4 (non-active) and 0.6 (from active) may work better, but left at 0.5 for simplicity of implementation and I think it's good enough?
- Border always uses the gradiented version. I did this for simplicity of implementation, but also think it looks better.
- Adjusted transform durations to feel better to me.

Let me know if you have any issues with any of these changes. Looks great otherwise.

bdach committed:
Remove unused using directive

bdach committed:
Merge branch 'master' into mod-overlay/panel

bdach commented:
Aside from one stray unused using (which I've removed) I have no issue with the latest changes, thanks.
